R/utilities.R

```diff
@@ -2236,4 +2236,175 @@ SparseSpearmanCor <- function(X, Y = NULL, cov = FALSE) {
+#' @importFrom future.apply future_lapply
+
+ExportGroupBW  <- function(
+  object = NULL,
+  assay = "peaks",
+  group.by = "seurat_clusters",
+  idents = c(0,2),
+  normMethod = "TSS.enrichment",
+  tileSize = 100,
+  minCells = 5,
+  cutoff = 4,
+  chromosome = "primary",
+  threads = NULL,
+  outdir = NULL,
+  verbose=TRUE
+  ){
+  
+    DefaultAssay(object) <- assay
+  
+    GroupsNames <- names(table(object@meta.data[,group.by])[table(object@meta.data[,group.by]) > minCells])
+    GroupsNames <- GroupsNames[GroupsNames %in% idents]
+       
+    #Column to normalized by
+    if (normMethod == 'ncells'){
+      normBy <- normMethod
+    } else{
+      normBy <- object@meta.data[, normMethod, drop=FALSE]
+    }
+    
+    #Get chromosome information
+    if(chromosome=="primary"){
+      prim_chr <- names(seqlengths(object)[!grepl("_alt|_fix|_random|chrUn", names(seqlengths(object)))])
+      seqlevels(object) <- prim_chr
+    }
+    availableChr <- names(seqlengths(object))
+    chromLengths <- seqlengths(object)
+    chromSizes <- GRanges(seqnames = availableChr, IRanges(start = rep(1, length(availableChr)), end = as.numeric(chromLengths)))
+
+    if (verbose) {
+      message("Creation of tiles")
+    }
+    
+    #Create tiles for each chromosome, from GenomicRanges
+    tiles <- unlist(slidingWindows(chromSizes, width = tileSize, step = tileSize))
+    
+    if (verbose) {
+      message("Bigwig files generation at ", outdir)
+    }
+    
+    #Set number of thread in future
+    plan("multicore", workers = threads)
+    
+    #Run the creation of bigwig for each cellgroups
+    covFiles <- future_lapply(GroupsNames, CreateBWGroup, availableChr, chromLengths, tiles, normBy, tileSize, normMethod, cutoff, outdir)
+
+    covFiles
+
+  }
+
+# Helper function for ExportGroupBW
+# @param groupNamei The group to be exported
+# @param availableChr Chromosomes to be processed
+# @param chromLengths Chromosomes lengths
+# @param tiles The tiles object
+# @param normBy A vector of values to normalized the cells by
+# or ncells as number of cells normalization
+# @param tileSize The size of the tiles in the bigwig file
+# @param normMethod Normalization method for the biwig files
+# It can be the name of any quantitative column in the @meta.data
+# or ncells as the number of cells
+# @param cutoff The maximum number of fragment in a given tile
+# @param outdir The output directory for bigwig file
+# Also used as output directory for SlitFragment function
+#' @importFrom rtracklayer import export.bw
+#' @importFrom GenomicRanges seqnames seqlengths GRanges coverage
+#' @importFrom BiocGenerics which
+#' @importFrom S4Vectors match
+#' @importFrom Matrix sparseMatrix rowSums
+CreateBWGroup <- function(groupNamei, availableChr, chromLengths, tiles, normBy, tileSize, normMethod, cutoff, outdir){
+
+  #Read the fragments file associated to the group
+  fragi <- rtracklayer::import(paste0(outdir, "/", groupNamei, ".bed"),format = "bed")
```

Comment:
also for this line, it will not work on windows where they use a different path delimiter. Need to use `.Platform$file.sep` to access the file separator used on the machine:

https://github.com/stuart-lab/signac/blob/af4142724b72574d957f7fe3d422ed5828ec3ad0/tests/testthat/test_fragments.R#L282